torch/csrc/jit/script/sugared_value.cpp

```diff
@@ -272,6 +290,15 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
     ctx_inputs.insert(ctx_inputs.end(), inputs.begin(), inputs.end());
     return FunctionValue(ret).call(loc, m, ctx_inputs, attributes, n_binders);
   }
+
+  if (auto class_type = getValue()->type()->cast<ClassType>()) {
+    auto self_nv = NamedValue(loc, "self", getValue());
```

Comment:
You are adding `self` to `inputs` here, only to slice it off in `callClassMethod` above. We seem to be piling pieces on top of each other here; it's not clear that callClassMethod should exist at all—could not be expressed as a composition of operations on sugared values? attr lookup (to get the `__call__` MethodValue), then calling the method with the provided arguments.

Reply:
an attr lookup above would give the error message:
> Tried to access nonexistent attribute `__call__`. Did you forget to initialize it in  _\_init\_\_()?:

The error message with `callClassMethod` is:
> Foo does not define a _\_call\_\_ method

which is a better error message. separately, looking up an attr and trying to call it is not actually correct - if you a class with an attribute `__call__` that is callable but is not a method than you can not invoke the class as a callable. 

Also - asking to inline a function that expresses a common pattern (it's invoked three times in this file) seems like a nit to me. Could we accept PRs when the only remaining comments are nits?

Reply:
This code does seem like a bandaid. Attribute lookup is how we always lookup methods. If the error message is bad here it is also bad for standard method lookup. It's fine to want to improve that but it is not ok to just fix it in this instance and add more code to maintain. I do not think it is a good habit to think these kinds of things are nits. By adding more pathways here rather than using existing ones, it leaves the code in worse shape than where we started.

Reply:
The error message is bad because it attr lookup in general does not have the additional context that we are looking for a method instead of another type of attribute (`did you forget to initialize it in __init__()?`. In other cases our error messages have been overly general and not applied to the actual problem at hand and it's led to confused users (`did you forget to add it to __constants__?`). 

This is specializing an error message when we know more information; in general should I consider that an anti-pattern?

Reply:
Attribute lookup is how we always lookup methods. This means the error for missing methods is always this bad. This code fixes it for the case of only call `__call__`. I am asking that is is either fixed for all methods or not fixed at all. The middle ground is more code to maintain and is still not good.

Reply:
I removed `callClassMethod`. It fixed it for all magic methods but anyways I updated the PR to remove the extra path.